### PR TITLE
Fixed #29363 -- Added SimpleTestCase.assertWarnsMessage().

### PR DESCRIPTION
**Pull request by ellmetha**

### CODE REVIEW

File: tests/cache/tests.py [hunk 1]
```diff
@@ -602,12 +601,8 @@ def func(key, *args):
         cache.key_func = func
 
         try:
-            with warnings.catch_warnings(record=True) as w:
-                warnings.simplefilter("always")
+            with self.assertWarnsRegex(CacheKeyWarning, expected_warning):
```
Comment (charettes) [inline tests/cache/tests.py]: A better approach might be to pass `'^%s$' % re.escape(expected_warning))` to `assertWarnsRegex` instead of manually escaping.

**Comment by ellmetha** [comment]
@charettes I've just added your first suggestion. Thanks for the review!

**Comment by timgraham** [comment]
Adding `assertWarnsMessage()` does look appealing considering the boilerplate of the current approach.

**Comment by ellmetha** [comment]
@timgraham Indeed. My last commit provides an ``assertWarnsMessage`` helper and updates the relevant unit tests. I've also updated some sections of the documentation related to testing tools and the coding style.

**Comment by carltongibson** [comment]
Slight puzzlement: 

> When used as a context manager, `assertWarns()` accepts the additional keyword argument `msg`.
> — https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarns

But seemingly no affect. ???

**Comment by charettes** [comment]
@carltongibson `msg` is used to format the raised exception message and not to test the raised warning. It's an analogous to `self.assertEqual(condition, msg)`.

https://github.com/python/cpython/blob/491bbedc209fea314a04cb3015da68fb0aa63238/Lib/unittest/case.py#L133-L135

**Comment by carltongibson** [comment]
@charettes: Super. Thank you.

File: docs/internals/contributing/writing-code/coding-style.txt [hunk 2]
```diff
@@ -67,6 +67,11 @@ Python style
   message. Use :meth:`~unittest.TestCase.assertRaisesRegex` only if you need
   regular expression matching.
 
+* In tests, use :meth:`~django.test.SimpleTestCase.assertWarnsMessage` instead
```
Comment (timgraham) [inline docs/internals/contributing/writing-code/coding-style.txt]: I'd combine this with the prefix note, "In tests, use assertRaisesMessage/assertWarnsMessage instead of assertRaises/assertWarns so ....

File: docs/topics/testing/tools.txt [hunk 3]
```diff
@@ -1362,6 +1364,22 @@ your test suite.
         with self.assertRaisesMessage(ValueError, 'invalid literal for int()'):
             int('a')
 
+.. method:: SimpleTestCase.assertWarnsMessage(expected_warning, expected_message, callable, *args, **kwargs)
+            SimpleTestCase.assertWarnsMessage(expected_warning, expected_message)
+
+    Asserts that execution of ``callable`` triggers ``expected_warning`` and
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: Rather than repeating two paragraphs, I think this could simply say, "Analogous to SimpleTestCase.assertRaisesMessage but for assertWarnsRegex instead of assertRaisesRegex."

Also, add a `.. versionadded:: 2.1` annotation.

File: django/test/testcases.py [hunk 4]
```diff
@@ -614,6 +614,36 @@ def assertRaisesMessage(self, expected_exception, expected_message, *args, **kwa
         with cm:
             callable_obj(*args, **kwargs)
 
+    @contextmanager
+    def _assert_warns_message_cm(self, expected_warning, expected_message):
+        with self.assertWarns(expected_warning) as cm:
+            yield cm
+        self.assertEqual(str(cm.warning), expected_message)
```
Comment (timgraham) [inline django/test/testcases.py]: There's a slight difference from assertRaisesMessage here in that this uses assertEqual instead of assertIn. Is it intentional? Not that I necessarily disagree with the decision. I haven't looked at the reason for using assertIn in the first place. In either case, either the code should be changed or the docstring should be updated.

File: tests/pagination/tests.py [hunk 5]
```diff
@@ -374,11 +368,11 @@ def test_paginating_unordered_object_list_raises_warning(self):
...
-            Paginator(object_list, 5)
-        self.assertEqual(len(warns), 1)
-        self.assertEqual(str(warns[0].message), (
+
+        msg = (
             "Pagination may yield inconsistent results with an unordered "
             "object_list: {!r}.".format(object_list)
-        ))
+        )
+
```
Comment (timgraham) [inline tests/pagination/tests.py]: remove extra blank line

File: tests/get_earliest_or_latest/tests.py [hunk 6]
```diff
@@ -169,16 +169,11 @@ def test_latest_manual(self):
...
-
-        self.assertEqual(len(warns), 1)
-        self.assertEqual(
-            str(warns[0].message),
+        msg = (
             'The field_name keyword argument to earliest() and latest() '
-            'is deprecated in favor of passing positional arguments.',
+            'is deprecated in favor of passing positional arguments.'
         )
+        self.assertWarnsMessage(RemovedInDjango30Warning, msg, Person.objects.latest, field_name='birthday')
```
Comment (timgraham) [inline tests/get_earliest_or_latest/tests.py]: Use the context manager version, please.

File: tests/fixtures/tests.py [hunk 7]
```diff
@@ -495,24 +495,16 @@ def test_dumpdata_proxy_without_concrete(self):
...
-            "fixtures.ProxySpy is a proxy model and won't be serialized."
-        )
 
     def test_dumpdata_proxy_with_concrete(self):
         """
         A warning isn't displayed if a proxy model is dumped with its concrete
         parent.
         """
         spy = ProxySpy.objects.create(name='Paul')
-
```
Comment (timgraham) [inline tests/fixtures/tests.py]: please remove the unrelated change

File: tests/deprecation/tests.py [hunk 8]
```diff
@@ -169,8 +150,6 @@ class Manager(metaclass=DeprecationInstanceCheck):
             alternative = 'fake.path.Foo'
             deprecation_warning = RemovedInNextVersionWarning
 
-        msg = '`Manager` is deprecated, use `fake.path.Foo` instead.'
-        with warnings.catch_warnings():
-            warnings.simplefilter('error', category=RemovedInNextVersionWarning)
-            with self.assertRaisesMessage(RemovedInNextVersionWarning, msg):
-                isinstance(object, Manager)
+        warning_msg = '`Manager` is deprecated, use `fake.path.Foo` instead.'
```
Comment (timgraham) [inline tests/deprecation/tests.py]: No need to change `msg` to `warning_msg`.

File: tests/backends/postgresql/tests.py [hunk 9]
```diff
@@ -23,21 +22,17 @@ def mocked_connect(self):
         self.assertIsNone(nodb_conn.settings_dict['NAME'])
 
         # Now assume the 'postgres' db isn't available
-        with warnings.catch_warnings(record=True) as w:
+        with self.assertWarns(RuntimeWarning):
```
Comment (timgraham) [inline tests/backends/postgresql/tests.py]: Verify the message as well, please.

File: django/test/testcases.py (hunk 4 above)
Comment (ellmetha) [reply django/test/testcases.py]: It seemed more appropriate to do so this way. I'll update the method to ensure that the behavior of`assertWarnsMessage` is similar to `assertRaisesMessage`. 😉

**Comment by ellmetha** [comment]
Thanks for the review @timgraham. I've made the proposed changes.
